[Chng Zi Hao] iP

zihaooo9 opened the pull request:
### Duke 
> “He who is not courageous enough to take risks will accomplish nothing in life.” – Muhammad Ali ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,

- Text-based
- Easy to learn
- ~Fast~ _SUPER FAST_ to use

All you need to do is,

1. Download it from [here](https://github.com/zihaooo9/ip).
2. Double-click it.
3. Add your tasks.
4. Let it manage your task for you 😉 

And its **FREE**!

Features: 

- [x] Managing task
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

##
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

zihaooo9 committed:
Completed Duke Level-1: Greet, Echo, Exit

zihaooo9 committed:
Completed Level 2. Add, List

zihaooo9 committed:
Completed Level 3. Mark as Done

zihaooo9 committed:
Fixed list limit in Level-3

zihaooo9 committed:
Completed Level 4. ToDos, Events, Deadlines.

zihaooo9 committed:
Completed A-TextUiTesting: Automated Text UI Testing

zihaooo9 committed:
Completed Level-5: Handle Errors

zihaooo9 committed:
Completed Level 6. Delete

zihaooo9 committed:
Implemented coding standards

zihaooo9 committed:
Completed Level 7. Save

zihaooo9 committed:
Add DateTime functionality

zihaooo9 committed:
Revert "Implemented coding standards"

This reverts commit d4612dcc3922605512b252a0fc7a0900dfb32f6d.

zihaooo9 committed:
Revert "Completed Level 7. Save"

This reverts commit b0d02fae6ab5c3e50532cfb48fcec1a9343ad20c.

zihaooo9 committed:
Revert "Revert "Implemented coding standards""

This reverts commit e5be49aba95e65a3247d2bb289ffa33ebfc7957c.

zihaooo9 committed:
Add load and save functionality

zihaooo9 committed:
Merge branch 'branch-Level-7'

zihaooo9 committed:
Merge branch 'branch-Level-8'

zihaooo9 committed:
Refactor code to make code more OOP

zihaooo9 committed:
Organise classes into packages and add subclasses for DukeException

zihaooo9 committed:
Add JUnit tests

zihaooo9 committed:
Create and release Jar file

zihaooo9 committed:
Add JavaDoc comments

zihaooo9 committed:
Tweak code to comply with a coding standard

zihaooo9 committed:
Add find task by keyword functionality

zihaooo9 committed:
Merge branch 'branch-A-JavaDoc'

zihaooo9 committed:
Merge branch 'branch-A-CodingStandard'

zihaooo9 committed:
Merge branch 'branch-Level-9'

zihaooo9 committed:
Update help menu

zihaooo9 committed:
Refine code to coding standards

zihaooo9 committed:
Merge branch 'add-gradle-support'

zihaooo9 committed:
Add checkstyle to detect coding violations

zihaooo9 committed:
Merge branch 'branch-A-CheckStyle'

zihaooo9 committed:
Improve code to follow coding standards

zihaooo9 committed:
Add GUI for Duke

zihaooo9 committed:
Merge branch 'branch-A-GUI'

jason-ng-zq99 commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,118 @@
+ */
+public class Storage {
+    private static final String DIRECTORY = "./data/";
+    private static final String FILEPATH = DIRECTORY + "data.txt";
```
I think it will be nicer to name `FILEPATH` as `FILE_PATH`

jason-ng-zq99 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,55 @@
+     */
+    private String formatDateTime() {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM dd yyyy");
+        String date = dueDate.format(formatter);
+        return date + " " + dueTime.toString();
```
Great catch on the more advanced rule of avoiding unnecessary use of this with fields! Spotted this change at other places as well

jason-ng-zq99 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @Override
+    public boolean hasDate(LocalDate date) {
```
Similarly, a great catch on the naming of the boolean methods! Spotted similar changes elsewhere too :smile:

jason-ng-zq99 commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,69 @@
+            + "| | | | | | | |/ / _ \\\n"
+            + "| |_| | |_| |   <  __/\n"
+            + "|____/ \\__,_|_|\\_\\___|\n";
+    private static final String WELCOMEMESSAGE = String.format("%s\n%s\n%s\n%s", DIVIDER, LOGO,
```
I think it will be nicer to name ```WELCOMEMESSAGE``` as ```WELCOME_MESSAGE```, spotted similar names somewhere else in this class as well

jason-ng-zq99 commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,246 @@
+     */
+    public Command parse(String input) throws DukeException {
+        Instruction instruction = Instruction.valueOfLabel(input.split(" ")[0]);
+        switch (instruction) {
+        case LIST:
+            return new ListCommand();
+        case DONE:
+            int doneIndex = extractIndex(input, Instruction.DONE);
+            return new DoneCommand(doneIndex);
+        case HELP:
+            return new HelpCommand();
+        case TODO:
+            String toDoInfo = extractInfo(input, Instruction.TODO);
+            ToDo toDo = new ToDo(toDoInfo);
+            return new AddCommand(toDo);
+        case DEADLINE:
+            String[] deadlineInfo = extractDeadline(extractInfo(input, Instruction.DEADLINE));
+            String[] deadlineDateTime = deadlineDateTimeSplit(deadlineInfo[1]);
+            Deadline deadline = new Deadline(deadlineInfo[0], parseDate(deadlineDateTime[0]),
+                    parseTime(deadlineDateTime[1]));
+            return new AddCommand(deadline);
+        case EVENT:
+            String[] eventInfo = extractEvent(extractInfo(input, Instruction.EVENT));
+            String[] eventDateTime = eventDateTimeSplit(eventInfo[1]);
+            Event event = new Event(eventInfo[0], parseDate(eventDateTime[0]), parseTime(eventDateTime[1]),
+                    parseTime(eventDateTime[2]));
+            return new AddCommand(event);
+        case DELETE:
+            int deleteIndex = extractIndex(input, Instruction.DELETE);
+            return new DeleteCommand(deleteIndex);
+        case FILTER:
+            String filterDate = extractInfo(input, Instruction.FILTER);
+            return new FilterCommand(parseDate(filterDate));
+        case FIND:
+            return new FindCommand(extractInfo(input, Instruction.FIND));
+        case BYE:
+            return new ExitCommand();
+        default:
+            return new InvalidCommand();
```
I believe it's better to add the ```//Fallthrough``` comment whenever there is a ```case``` statement without a ```break``` statement. 
It helps to make clear that this is intentional and not by accident :smile: 
Example:
```javascript
switch (condition) {
case ABC:
    statements;
    // Fallthrough
case DEF:
    statements;
    break;
```
Saw this from the [textbook](https://se-education.org/guides/conventions/java/intermediate.html#layout).

jason-ng-zq99 reviewed (commented):
I think your code is great aside from the small issues I pointed out regarding coding violations!
All the best for the remaining development of Duke! :smile:

ChuangZheQuan commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * Deletes task from the task list.
+     *
+     * @param taskList   TaskList of Duke.
+     * @param ui      The user interface.
```
Maybe you can standardise the number of spaces?

ChuangZheQuan commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,38 @@
+ *
+ * @author Chng Zi Hao
+ */
+public class FindCommand extends Command {
```
Maybe can be renamed to ```CommandFinder``` which is more suitable to be a Class name?

ChuangZheQuan commented on src/main/java/duke/command/HelpCommand.java:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @author Chng Zi Hao
+ */
+public class HelpCommand extends Command {
```
Maybe can be renamed to CommandHelper which is more suitable to be a Class name?

ChuangZheQuan commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * Change status of task such that is is mark as done.
+     */
+    public void markDone() {
```
Maybe rename to```markAsDone```?

ChuangZheQuan reviewed (commented):
Good job! Just made some minor naming suggestions!

weiquu commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,72 @@
+ * @author Chng Zi Hao
+ */
+public class Duke {
+    private Storage storage;
+    private TaskList taskList;
+    private final Parser parser;
+    private final Ui ui;
```
Should these all be final if they won't be changed? Also since they will be constants, perhaps you could capitalise their names to indicate that?

weiquu commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,72 @@
+//                ui.formatPrint(e.getMessage());
+//            }
+//        }
+//    }
```
Do you still require this code? If not, you could consider removing it (instead of leaving it in as a chunk of commented code).

weiquu commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * Marks specified task as done.
+     *
+     * @param taskList   TaskList of Duke.
+     * @param ui      The user interface.
+     * @param storage Storage for Duke.
+     * @return Message to be shown to user.
+     * @throws DukeException If taskIndex is invalid.
```
As mentioned by Zhe Quan, you could consider standardising the number of spaces here.

weiquu commented on src/main/java/duke/exception/DukeInvalidCommandException.java:
```diff
@@ -0,0 +1,16 @@
+ * @author Chng Zi Hao
+ */
+public class DukeInvalidCommandException extends DukeException {
+
```
Just nitpicking here, but perhaps removing the empty line here to standardise formatting with the rest of the classes would be good? Same comment applies to the rest of the classes that extend `DukeException`.

weiquu commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,55 @@
+ * @author Chng Zi Hao
+ */
+public class Deadline extends Task {
+    private LocalDate dueDate;
+    private LocalTime dueTime;
```
Consider adding JavaDoc comments for these if possible?

weiquu commented on src/main/java/duke/util/Instruction.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * Instruction represents the set of instructions that Duke recognises.
+ */
+public enum Instruction {
```
I liked the use of enumerations here to have a concise view of which instructions Duke can take!

weiquu commented on src/main/java/duke/util/Instruction.java:
```diff
@@ -0,0 +1,39 @@
+     * @param label command that user input.
+     * @return a Command that corresponds to user input. Returns INVALID if input does not match any Commands.
+     */
+    public static Instruction valueOfLabel(String label) {
```
Maybe rename this to `getValueofLabel`? To be in line with the coding standard of having method names be verbs.

weiquu commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,246 @@
+    public static LocalDate parseDate(String date) throws DukeInvalidDateException {
+        try {
+            // Reuse regex from
+            // https://www.javacodeexamples.com/java-regular-expression-validate-date-example-regex/1504
```
Good crediting of sources!

weiquu commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,246 @@
+     * @return Array of String with date and time separated.
+     * @throws DukeMissingDateTimeException If input does not contain time or date.
+     */
+    public static String[] deadlineDateTimeSplit(String dateTime) throws DukeMissingDateTimeException {
```
Maybe place the verb at the front of the name, like `splitDeadlineDateTime`? This would then follow the standard set by the rest of the method names. Same comment for the method below!

weiquu commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,150 @@
+        if (list.size() == 0) {
+            return "List is empty. Add something to the list!\n";
+        }
+        String ls = "Here are the tasks in your list:\n";
```
Might want to use a more descriptive name here? But I also get that it's only for a short section of the method.

weiquu commented on src/test/java/duke/util/TaskListTest.java:
```diff
@@ -0,0 +1,107 @@
+    // @author Jonathan Cook
+    // Reused from https://www.baeldung.com/java-testing-system-out-println
+    // with minor modifications
+    private final PrintStream standardOut = System.out;
+    private final ByteArrayOutputStream outputStreamCaptor = new ByteArrayOutputStream();
```
You might want to consider changing the variable names to `STANDARD_OUT` and `OUTPUT_STREAM_CAPTOR`? Since they are declared as being final.

weiquu reviewed (commented):
Overall, I think this was a really great job! There were a few minor coding standard violations, such as constant names not being capitalised (and separated by underscores) as well as some methods that could have had better names. That being said, the code was very readable and I liked the implementation of having enumerations and the use of switch statements. Good job and also was nice to see credit being given in the comments!

zihaooo9 committed:
Add assertions

zihaooo9 committed:
Add assertions

There are several assumptions made in some methods, which can lead to some bugs in the code.

Having assertions allows testing of the correctness of any assumptions that have been made in the program.

Let's insert assertions for Internal Invariants, Control-Flow Invariants, Preconditions, Postconditions, and Class Invariants.

zihaooo9 committed:
Merge commit '3d5859691717e38af63c861d9b925377ac412f78' into branch-A-Assertions

zihaooo9 committed:
Improve code quality

There are several methods that do not adhere to the code quality
standards specified in the textbook.

Improving the code quality will significantly increase readability and
understandability of the code. This helps ourselves and other developers
who wish to use the code in the future.

Let's,
* refactor the code to make it more readable

zihaooo9 committed:
Merge pull request #1 from zihaooo9/branch-A-Assertions

Branch a assertions

zihaooo9 committed:
Merge branch 'master' of github.com:zihaooo9/ip

zihaooo9 committed:
Merge branch 'master' into branch-A-CodeQuality

zihaooo9 committed:
Merge pull request #2 from zihaooo9/branch-A-CodeQuality

Improve code quality

zihaooo9 committed:
Add help functionality

The program may be difficult to use for some new users and these users
may require some assistance to learn how to use the CLI Chat Bot.

Having a help menu and mini tutorial for first-time users helps these
users orientate themselves to the app, making it more user-friendly.

Let's,
* Add a help menu
* Tutorial for first-time users

zihaooo9 committed:
Merge branch 'branch-C-Help'

zihaooo9 committed:
Improve GUI

zihaooo9 committed:
Add User Guide

zihaooo9 committed:
Set theme jekyll-theme-cayman

zihaooo9 committed:
Set theme jekyll-theme-tactile

zihaooo9 committed:
Update _config.yml

zihaooo9 committed:
Update JavaDocs

zihaooo9 committed:
Set theme jekyll-theme-cayman

zihaooo9 committed:
update build.gradle
